### PR TITLE
Auto-upcast and reshape mask

### PR DESCRIPTION
**Pull request by ragulpr**
On GPU it runs as;

```
y = torch.randn(512,128,512, requires_grad=True,device='cuda')
dropout = TailDropout().cuda()
optimizer = torch.optim.SGD((y,), lr=0.1)
# dropout.compile()
for _ in range(1000):
  z = dropout(y)
  loss = z.sum()
  loss.backward()
  optimizer.step()
  optimizer.zero_grad()
```
```
%%timeit
for _ in range(1000):
  z = dropout(y)
  loss = z.sum()
  loss.backward()
  optimizer.step()
  optimizer.zero_grad()

# 4.03 s ± 1.59 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)# new
# 4.04 s ± 743 µs per loop (mean ± std. dev. of 7 runs, 1 loop each) # new
# 5.09 s ± 843 µs per loop (mean ± std. dev. of 7 runs, 1 loop each) # Compiled


# 4.05 s ± 1.25 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # old
# 5.08 s ± 715 µs per loop (mean ± std. dev. of 7 runs, 1 loop each) # Old compiled

```

### CODE REVIEW

File: taildropout.py [hunk 1]
```diff
@@ -135,15 +135,16 @@ def forward(self, input: Tensor) -> Tensor:
...
             prob = self.cdf(linspace, self.scale * n_features)
 
+            prob_shape = replace_w_ones_except(input.shape, self.dropout_dim) #[1,1,..,n_features]
+            prob = prob.reshape(prob_shape)
+
             mask_shape = replace_w_ones_except(input.shape, self.batch_dim)
             uniform = torch.rand(mask_shape, device=device, dtype=type_out) # [n_batch,1,1] if input 3d
             mask = prob < uniform         # 43% of cpu cumtime                [n_batch,1,n_features] 
-            mask = mask.type(type_out)    # 30% of cpu cumtime
+            # mask = mask.type(type_out)    # 30% of cpu cumtime
```
Comment (ragulpr) [inline taildropout.py:147]: Keep comment for now as it may have been explicit to get the timing in the first place. Timing probably needs an update too.
